### PR TITLE
adding redis as indexer DB layer for de-duplication detection

### PR DESCRIPTION
**Pull request by harsh-mahajan-yral**
resolves https://github.com/dolr-ai/product-roadmap/issues/347

### CODE REVIEW

File: src/qstash/duplicate.rs [hunk 1]
```diff
@@ -239,69 +239,61 @@ impl<'a> VideoHashDuplication<'a> {
...
-                .await?;
-
-                log::info!(
-                    "Duplicate video detected: video_id [{}] is similar to parent_video_id [{}] (score: {})",
-                    video_id,
-                    match_details.video_id,
-                    match_details.similarity_percentage
-                );
+            // Get the original video ID associated with this hash
+            let parent_video_id: String = redis::cmd("GET")
```
Comment (komal-sai-yral) [inline src/qstash/duplicate.rs:263]: lets use HSET HGET instead - https://www.dragonflydb.io/docs/command-reference/hashes/hset (dragonfly has better redis docs than redis :P ) . in near future we might have to store hash reference counts and other related things 

also can we use typed way of doing this instead . talking abt - https://docs.rs/redis/latest/redis/struct.Cmd.html#method.hset

File: src/qstash/duplicate.rs [hunk 2]
```diff
@@ -239,69 +239,61 @@ impl<'a> VideoHashDuplication<'a> {
...
+        let is_duplicate: bool = redis::cmd("EXISTS")
+            .arg(&video_hash.hash)
+            .query_async(&mut con)
+            .await
+            .map_err(|e| anyhow::anyhow!("Failed to query Redis: {}", e))?;
 
         if is_duplicate {
-            // A similar video was found - record as duplicate
-            if let Some(match_details) = indexer_response.match_details {
-                self.store_duplicate_video(
```
Comment (komal-sai-yral) [inline src/qstash/duplicate.rs:275]: do cleanup unused methods

File: src/qstash/duplicate.rs [hunk 3]
```diff
@@ -427,4 +419,52 @@ impl<'a> VideoHashDuplication<'a> {
...
     }
+
+    async fn store_duplicate_video_exact(
+        &self,
+        video_id: &str,
+        hash: &str,
+        parent_video_id: &str,
+        publisher_data: &VideoPublisherData,
+    ) -> Result<(), anyhow::Error> {
+        let bigquery_client = app_state::init_bigquery_client().await;
```
Comment (komal-sai-yral) [inline src/qstash/duplicate.rs:430]: reuse already present client

File: src/qstash/duplicate.rs [hunk 4]
```diff
@@ -427,4 +419,52 @@ impl<'a> VideoHashDuplication<'a> {
...
+            ..Default::default()
+        };
+
+        log::info!(
+            "Storing exact duplicate video in duplicate_videos: video_id [{}], parent_video_id [{}], score=100.0",
+            video_id,
+            parent_video_id
+        );
+
+        bigquery_client
```
Comment (komal-sai-yral) [inline src/qstash/duplicate.rs:463]: lets start reading the result, parsing and printing to highlight any issues here. also move away from job for inserts. check this - https://github.com/dolr-ai/off-chain-agent/blob/242dfcfb09999cdae9424655c3eaa92cabeee8b4/src/posts/utils.rs#L23

File: src/qstash/duplicate.rs [hunk 5]
```diff
@@ -239,69 +239,61 @@ impl<'a> VideoHashDuplication<'a> {
...
-        if !response.status().is_success() {
-            let status = response.status();
-            let error_text = response.text().await.unwrap_or_default();
-            return Err(anyhow::anyhow!(
-                "VideoHash Indexer API failed: {} - {}",
-                status,
-                error_text
-            ));
-        }
+        let client = redis::Client::open(redis_url)
```
Comment (komal-sai-yral) [inline src/qstash/duplicate.rs:246]: use app_state for init as you mentioned

**Review by komal-sai-yral — COMMENTED** [review]
do document the backfill and deployment plan so that we can review

**Comment by harsh-mahajan-yral** [comment]
complete flow for porting from videohash_indexer to redis

- change the interface for duplication check in offchain from videohash_indexer to redis
- get this PR approve but not merge
- simultaneously we will backfill the redis DB with videohashs and video_ids, with big query table
- when the backfill completes, we will merge the redis interface PR
- then again is some entries are left we can just call the apis backfill api to sync redis table

File: src/qstash/duplicate.rs [hunk 6]
```diff
@@ -235,76 +131,71 @@ impl<'a> VideoHashDuplication<'a> {
             .await
             .map_err(|e| anyhow::anyhow!("Failed to generate videohash: {}", e))?;
 
-        // Store the original hash regardless of duplication status
-        self.store_videohash_original(video_id, &video_hash.hash)
-            .await?;
+        // Get BigQuery client once
+        let bigquery_client = app_state::init_bigquery_client().await;
```
[the remark this answers is not in the document]
Comment (harsh-mahajan-yral) [reply src/qstash/duplicate.rs:135]: missed those

**Comment by komal-sai-yral** [comment]
> complete flow for porting from videohash_indexer to redis
> 
>     * change the interface for duplication check in offchain from videohash_indexer to redis
> 
>     * get this PR approve but not merge
> 
>     * simultaneously we will backfill the redis DB with videohashs and video_ids, with big query table
> 
>     * when the backfill completes, we will merge the redis interface PR
> 
>     * then again is some entries are left we can just call the apis backfill api to sync redis table

Sounds good. Couple of things:
- before this, the first step would be to clean video_unique and add missing values. GROUP BY videohash_original FILTER video_deleted. ensure only one member of a group exists in video_unique, if missing add any random member should be fine . ofc with proper backups, experimenting on clone and only then applying on prod table
- get rid of duplicate_videos completely since not being used. lets not update in code as well. will reduce one latency step
- testing the redis backfill and new flow in preview
- then the above steps of redis update and new flow merge

File: src/qstash/duplicate.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (komal-sai-yral) [reply src/qstash/duplicate.rs:121]: actually no, its reverse.

File: src/qstash/duplicate.rs [hunk 7]
```diff
@@ -235,76 +120,64 @@ impl<'a> VideoHashDuplication<'a> {
...
-                    similarity_percentage: match_details.similarity_percentage,
-                    exact_duplicate,
-                    publisher_canister_id: publisher_data.canister_id.clone(),
-                    publisher_principal: publisher_data.publisher_principal.clone(),
-                    post_id: publisher_data.post_id,
-                    timestamp: chrono::Utc::now().to_rfc3339(),
-                };
-            }
+            // Store duplicate using the same BigQuery client
+            self.store_duplicate_video_exact(
```
Comment (komal-sai-yral) [inline src/qstash/duplicate.rs:157]: lets not store duplicate_videos. not being used anywhere

File: src/qstash/duplicate.rs [hunk 8]
```diff
@@ -370,61 +273,88 @@ impl<'a> VideoHashDuplication<'a> {
...
+                ));
+            }
+        }
+
         Ok(())
     }
 
-    async fn store_duplicate_video(
+    // Updated method for storing exact duplicates
+    async fn store_duplicate_video_exact(
```
Comment (komal-sai-yral) [inline src/qstash/duplicate.rs:299]: don't need this method

**Review by komal-sai-yral — COMMENTED** [review]
looks good. minor cleanup. also pls test e2e uploading a unique video, duplicate and check values in redis, and 2 BQ tables

File: src/qstash/duplicate.rs [hunk 9]
```diff
@@ -235,76 +120,40 @@ impl<'a> VideoHashDuplication<'a> {
...
-            }
-        } else {
-            self.store_unique_video(video_id, &video_hash.hash).await?;
-            log::info!("Unique video recorded: video_id [{}]", video_id);
-        }
+        // This is a unique video
+        log::info!("Unique video recorded: video_id [{}]", video_id);
+
+        // Store as unique using the same BigQuery client
+        self.store_unique_video(video_id, &video_hash.hash, bigquery_client)
```
Comment (komal-sai-yral) [inline src/qstash/duplicate.rs:147]: we need to do this only if !duplicate right ?

**Review by komal-sai-yral — CHANGES_REQUESTED** [review]

File: src/qstash/duplicate.rs (hunk 9 above)
Comment (harsh-mahajan-yral) [reply src/qstash/duplicate.rs:147]: yes

Comment (harsh-mahajan-yral) [reply src/qstash/duplicate.rs:147]: ok wait smjha

**Review by komal-sai-yral — COMMENTED** [review]
looks good. do test and confirm. will approve
